### ISSUE TITLE
💅 lint/suspicious/useIterableCallbackReturn may incorrectly report

### ISSUE THREAD

**Issue by Jiralite** [biomejs/biome#7229]
### Environment information

```bash
CLI:
  Version:                      2.2.0
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.5.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.14.0

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         /Users/jiralite/Documents/GitHub/thatskyapplication/thatskyapplication/biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noForEach
    complexity/noImportantStyles
    complexity/noStaticOnlyClass
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    complexity/useSimplifiedLogicExpression
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUndeclaredDependencies
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useGraphqlNamedOperations
    correctness/useIsNan
    correctness/useJsxKeyInIterable
    correctness/useParseIntRadix
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/noFloatingPromises
    performance/noAccumulatingSpread
    performance/noBarrelFile
    performance/noDelete
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noDescendingSpecificity
    style/noInferrableTypes
    style/noNegationElse
    style/noParameterAssign
    style/noUnusedTemplateLiteral
    style/noUselessElse
    style/noYodaExpression
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useBlockStatements
    style/useCollapsedElseIf
    style/useCollapsedIf
    style/useConsistentArrayType
    style/useConsistentCurlyBraces
    style/useConsistentMemberAccessibility
    style/useConst
    style/useDefaultParameterLast
    style/useDeprecatedReason
    style/useEnumInitializers
    style/useExplicitLengthCheck
    style/useExponentiationOperator
    style/useExportType
    style/useForOf
    style/useFragmentSyntax
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useSelfClosingElements
    style/useShorthandAssign
    style/useShorthandFunctionType
    style/useSingleVarDeclarator
    style/useTemplate
    suspicious/noAlert
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noBiomeFirstException
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstantBinaryExpressions
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnknownAtRules
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessEscapeInString
    suspicious/noUselessRegexBackrefs
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useIterableCallbackReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

useIterableCallbackReturn

### Playground link

https://biomejs.dev/playground/?lineWidth=120&lintRules=useIterableCallbackReturn&code=ZQBuAHUAbQAgAFQAeQBwAGUAIAB7AAoACQBIAGUAbABsAG8AIAA9ACAAMAAsAAoAfQAKAAoAYwBvAG4AcwB0ACAAQQBSAFIAQQBZACAAPQAgAFsACgAJAHsACgAJAAkAdAB5AHAAZQA6ACAAVAB5AHAAZQAuAEgAZQBsAGwAbwAsAAoACQAJAG0AZQBzAHMAYQBnAGUAOgAgACIASABlAGwAbABvACEAIgAsAAoACQB9ACwACgBdACAAYQBzACAAYwBvAG4AcwB0ADsACgAKAEEAUgBSAEEAWQAuAHIAZQBkAHUAYwBlACgAKABtAGEAcAAsACAAdABoAHIAZQBhAGQAKQAgAD0APgAgAHsACgAJAHMAdwBpAHQAYwBoACAAKAB0AGgAcgBlAGEAZAAuAHQAeQBwAGUAKQAgAHsACgAJAAkAYwBhAHMAZQAgAFQAeQBwAGUALgBIAGUAbABsAG8AOgAKAAkACQAJAHIAZQB0AHUAcgBuACAAbQBhAHAALgBzAGUAdAAoAFQAeQBwAGUALgBIAGUAbABsAG8ALAAgAHQAaAByAGUAYQBkAC4AbQBlAHMAcwBhAGcAZQApADsACgAJAH0ACgB9ACwAIABuAGUAdwAgAE0AYQBwADwAVAB5AHAAZQAsACAAcwB0AHIAaQBuAGcAPgAoACkAKQA7AAoA&language=ts

### Expected result

The rule should not report that it must always return a value... because it does.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by dyc3** [comment]
The rule looks like it's is working as intended. It doesn't take into account type information when evaluating, and the switch missing the `default` case makes this not always return a value.

**Comment by dyc3** [comment]
See also: https://github.com/typescript-eslint/typescript-eslint/issues/2841

**Comment by Profesor08** [comment]
There is simple playground: [useIterableCallbackReturn](https://biomejs.dev/playground/?lineWidth=120&lintRules=useIterableCallbackReturn&code=aQBuAHQAZQByAGYAYQBjAGUAIABGAHIAdQBpAHQAIAB7AAoAIAAgAHQAeQBwAGUAOgAgACIAYQBwAHAAbABlACIAIAB8ACAAIgBiAGEAbgBhAG4AYQAiADsACgB9AAoACgBmAHUAbgBjAHQAaQBvAG4AIAB0AGUAcwB0ACgAZgByAHUAaQB0AHMAOgAgAEYAcgB1AGkAdABbAF0AKQAgAHsACgAgACAAcgBlAHQAdQByAG4AIABmAHIAdQBpAHQAcwAuAG0AYQBwACgAZgByAHUAaQB0ACAAPQA%2BACAAewAKACAAIAAgACAAcwB3AGkAdABjAGgAKABmAHIAdQBpAHQALgB0AHkAcABlACkAIAB7AAoAIAAgACAAIAAgACAAYwBhAHMAZQAgACIAYQBwAHAAbABlACIAOgAgAHIAZQB0AHUAcgBuACAAIgBhACIAOwAKACAAIAAgACAAIAAgAGMAYQBzAGUAIAAiAGIAYQBuAGEAbgBhACIAOgAgAHIAZQB0AHUAcgBuACAAIgBiACIAOwAKACAAIAAgACAAfQAKACAAIAB9ACkAOwAKAH0ACgAKAGkAbgB0AGUAcgBmAGEAYwBlACAARgByAHUAaQB0AFcAaQB0AGgAQgB1AGcAIAB7AAoAIAAgAHQAeQBwAGUAOgAgACIAYQBwAHAAbABlACIAIAB8ACAAIgBiAGEAbgBhAG4AYQAiACAAfAAgACIAYgB1AGcAIgA7AAoAfQAKAAoAZgB1AG4AYwB0AGkAbwBuACAAdABlAHMAdABfAGIAdQBnACgAZgByAHUAaQB0AHMAOgAgAEYAcgB1AGkAdABXAGkAdABoAEIAdQBnAFsAXQApACAAewAKACAAIAByAGUAdAB1AHIAbgAgAGYAcgB1AGkAdABzAC4AbQBhAHAAKABmAHIAdQBpAHQAIAA9AD4AIAB7AAoAIAAgACAAIABzAHcAaQB0AGMAaAAoAGYAcgB1AGkAdAAuAHQAeQBwAGUAKQAgAHsACgAgACAAIAAgACAAIABjAGEAcwBlACAAIgBhAHAAcABsAGUAIgA6ACAAcgBlAHQAdQByAG4AIAAiAGEAIgA7AAoAIAAgACAAIAAgACAAYwBhAHMAZQAgACIAYgBhAG4AYQBuAGEAIgA6ACAAcgBlAHQAdQByAG4AIAAiAGIAIgA7AAoAIAAgACAAIAAgACAAZABlAGYAYQB1AGwAdAA6ACAAcgBlAHQAdQByAG4AIAAiAGEAbABsACAAZgBpAG4AZQAiADsACgAgACAAIAAgAH0ACgAgACAAfQApADsACgB9AA%3D%3D&language=ts)

This rule ignores type information. Adding some `default` to switch will kill type sefety, because when the `Fruit.type` will expand with new values, there will be no alerts, and app will have bug in production.

**Comment by dyc3** [comment]
You could throw an error in the default case to catch the bug.

```ts
function test(fruits: Fruit[]) {
  return fruits.map(fruit => {
    switch(fruit.type) {
      case "apple": return "a";
      case "banana": return "b";
      default: throw new Error("hit bug")
    }
  });
}
```

**Comment by Profesor08** [comment]
This approach makes tools for static code analysis useless, because in any case, this puts us to validate everything in runtime. Non-user input must be validated for no reason; this is ridiculous. And when app is shipped, we get crashes or partially disabled functionality, because all tools had no alerts on build time. It is like using `any` or `// @ts-ignore` everywhere.

**Comment by ematipico** [comment]
Closing as the rule works as intended. You could use https://biomejs.dev/linter/rules/use-exhaustive-switch-cases/javascript/
